File: _mcp/dedao-vault-mcp/dedao_vault_mcp/columns.py
from pathlib import Path

import yaml

_SUMMARY_KEYS = ("id", "name", "format", "language", "topic_mode")


def _columns_path(creation_root) -> Path:
    return Path(creation_root) / "_栏目.yaml"


def load_columns(creation_root):
    path = _columns_path(creation_root)
    if not path.exists():
        return {"error": "columns_file_missing", "path": str(path)}
    try:
        data = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    except yaml.YAMLError as e:
        return {"error": "yaml_error", "detail": str(e)}
    cols = data.get("columns") or []
    return [c for c in cols if isinstance(c, dict) and "id" in c]


def list_columns(creation_root):
    cols = load_columns(creation_root)
    if isinstance(cols, dict):
        return cols
    return [{k: c.get(k) for k in _SUMMARY_KEYS} for c in cols]


def get_column(creation_root, id: str):
    cols = load_columns(creation_root)
    if isinstance(cols, dict):
        return cols
    for c in cols:
        if c.get("id") == id:
            return c
    return {"error": "column_not_found", "available": [c.get("id") for c in cols]}

File: _mcp/dedao-vault-mcp/src/columns.ts
import * as path from "node:path";
import * as fs from "node:fs";
import * as YAML from "yaml";

const SUMMARY_KEYS = ["id", "name", "format", "language", "topic_mode"];

function columnsPath(creationRoot: string): string {
  return path.join(creationRoot, "_栏目.yaml");
}

export function loadColumns(creationRoot: string): any[] | Record<string, any> {
  const p = columnsPath(creationRoot);
  if (!fs.existsSync(p)) return { error: "columns_file_missing", path: p };
  let data: any;
  try {
    data = YAML.parse(fs.readFileSync(p, "utf-8")) ?? {};
  } catch (e) {
    return { error: "yaml_error", detail: String(e) };
  }
  const cols = Array.isArray(data.columns) ? data.columns : [];
  return cols.filter((c: any) => c && typeof c === "object" && "id" in c);
}

export function listColumns(creationRoot: string) {
  const cols = loadColumns(creationRoot);
  if (!Array.isArray(cols)) return cols;
  return cols.map((c) => Object.fromEntries(SUMMARY_KEYS.map((k) => [k, c[k] ?? null])));
}

export function getColumn(creationRoot: string, id: string) {
  const cols = loadColumns(creationRoot);
  if (!Array.isArray(cols)) return cols;
  const found = cols.find((c) => c.id === id);
  if (found) return found;
  return { error: "column_not_found", available: cols.map((c) => c.id) };
}
